All custom reporter events

just-boris opened the pull request:
This is upgraded version of @gronke's PR #8.

Main differences:

1. Do not create `Instance` utility class. Allure commons module itself can take care about suite structure management.
2. Use cheerio for better output xml assertions. It provides jQuery like selectors with is probably better to query data in results.
3. Simplify test helper class
4. Tests are generally rewritten to fix eslint warnings, no actual changes

TBD:

* [x] test against parallel runs
* [x] handle errors in response

gronke commented:
(Different to master) I was not able to run this against tests in production. Missing support for concurrent test instances was not an issue with the non-event driven solution. After switching to one instance only that problem was gone, but it stuck at a test with failing before task.

Feel free to squash my commits so that we can keep the commit history straight forward when adding more features on top of this.

just-boris commented:
Okay, I will add configuration with more browsers, to check, how does it work in parallel.

Is it enough to reproduce multithreading issue?

gronke commented:
I was using multiple instances of the same browser. As long as the tests are non-trivial they all fire the `suite:start` event before even the first test was even started. The Instances class' job was to deal with that and create each an Allure instance per running webdriver instance.

just-boris commented:
@gronke could you provide an example of your tests setup?
Because everything, that I am doing now, just make passing all tests in the project

just-boris commented:
@gronke fixed parallel issues, you can try this again

just-boris commented:
@gronke I fixed parallel issues, you can try this again

just-boris commented:
@gronke I fixed parallel issues, you can try this again

gronke commented:
@just-boris thanks, running all tests without a crash now.

Here're a few lines of the output which look like Allure lost track in hooks:
```
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/element is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/element/18/click is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/url is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/elements is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/elements is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/elements is not whithin any running test
POST /wd/hub/session/c353a905-6131-48db-aed5-92f530b59192/elements is not whithin any running test
```

It's a bit tricky, because the order of the events indicates that a test already finished before the after hook runs. When a failure occurs during step 5 or 6, the test itself was already finished.
1. `suite:start`
2. `test:start`
3. `test:pass`
4. `test:end`
5. `hook:start`
6. `hook:end` 

When generating the HTML output, Allure complains about empty attributes `stop=""` or 'status=""`. Tested this on current master without that issue. We should fix that before merging.

just-boris commented:
Thanks. I added this warning exactly to catch that issues. 

I think it is caused by some actions in afterEach hook. Will look this evening at it. I fixed something similar in mocha-allure-reporter, so it will not be hard to fix.

gronke commented:
[The solution in mocha-allure-reporter](https://github.com/allure-framework/mocha-allure-reporter/blob/master/index.js#L51-L55) will probably not work with the webdriver BaseReporter, because we don't know if a tests is followed by one or more hooks. The only thing that we know for sure is that the next `start:test` will be fired after all hooks were finished. That's exactly the solution I came up with in the other branch, but it messed up the complexity. At the end I don't see a chance to get around tracking the status of suites and tests in a data-structure.

just-boris commented on test/specs/screenshot.js:
```diff
@@ -2,70 +2,48 @@
-      
-      let screenshotFiles = helper.getResultFiles(['jpg', 'png'])
-      expect(screenshotFiles, 'no screenshot files attached').to.have.lengthOf(1)
+    xit('[not supported yet] can be taken in an "before all" hook', () => {
```
I have this test ignored because I couldn't find a simple solution for it.

Now there is a warning about command calls outside of running tests (It happens when you have some actions in before all). I tried to add them, but then you will always would have step about getting session, because it is also logged.

We need to find a way how to filter session commands in another way, rather than just matching with `/wd/hub/session` as well as it going now with screenshot logging

just-boris commented:
@gronke your issue was caused by `before` hook. See [my comment](https://github.com/webdriverio/wdio-allure-reporter/pull/17/files#r67063548) about ignored tests.

I will try to find some solution to solve it, if it is real blocker to update.

just-boris commented:
Well, I enabled the test and made it pass.
Now there is a special test-case named `"before all" hook`, if there was some commands to dump.
Also failures at `"after all"` are also logged into special test-case with corresponding name

gronke commented:
Can you please patch this into the `package.json`, so that we support older node versions too

```diff
diff --git a/package.json b/package.json
index 2ecdc6d..b46e416 100644
--- a/package.json
+++ b/package.json
@@ -5,7 +5,7 @@
   "main": "build/reporter.js",
   "scripts": {
     "build": "grunt",
-    "test": "mocha test/specs --timeout 10000",
+    "test": "node --harmony_rest_parameters ./node_modules/grunt-cli/bin/grunt test",
     "prepublish": "npm prune && npm run build"
   },
   "repository": {
```

gronke commented:
This diff fixes the `outputDir` option. BaseReporter [flattens the reporterOptions when reporter and option name match](https://github.com/webdriverio/webdriverio/blob/0d5d2c57aace93e747eefd7a0fea9d2c26330ec2/lib/launcher.js#L83)

```diff
diff --git a/lib/reporter.js b/lib/reporter.js
index 7bb277b..1b09dc8 100644
--- a/lib/reporter.js
+++ b/lib/reporter.js
@@ -20,7 +20,9 @@ class AllureReporter extends events.EventEmitter {
 
         this.baseReporter = baseReporter
         this.config = config
-        this.options = options.allure || {}
+        this.options = {
+            outputDir: options.outputDir
+        }
         this.allures = {}
 
         const { epilogue } = this.baseReporter
```

gronke commented:
When running more complex test suites this error is thrown
```
/Projects/iDiscover/frontend/node_modules/webdriverio/build/lib/utils/BaseReporter.js:322
                        throw _iteratorError;
                        ^

TypeError: Cannot read property 'addStep' of undefined
    at Allure.startStep (/Projects/gronke/wdio-allure-reporter/node_modules/allure-js-commons/index.js:57:16)
    at AllureReporter.<anonymous> (/Projects/gronke/wdio-allure-reporter/build/reporter.js:117:20)
    at emitOne (events.js:90:13)
    at AllureReporter.emit (events.js:182:7)
    at BaseReporter.handleEvent (/Projects/iDiscover/frontend/node_modules/webdriverio/build/lib/utils/BaseReporter.js:310:35)
    at Launcher.messageHandler (/Projects/iDiscover/frontend/node_modules/webdriverio/build/lib/launcher.js:471:28)
    at emitTwo (events.js:100:13)
    at ChildProcess.emit (events.js:185:7)
    at handleMessage (internal/child_process.js:718:10)
    at Pipe.channel.onread (internal/child_process.js:444:11)
```

The line numbers will probably not match exactly because of the suggested changes from my last two comments. The origin line of this error is https://github.com/just-boris/wdio-allure-reporter/blob/ca4e5ec573a09a7ee3c7043e62b677849d506f0e/lib/reporter.js#L78 and it seems to be caused by commands executed out of the scope of a running test. When [allure-js-commons then tries add a step to a non-existing parent](https://github.com/allure-framework/allure-js-commons/blob/master/index.js#L57) it crashes.

Will try to extract some integration test from that, but I think this should not block us from merging and fixing in a later PR. @just-boris What do you think?

just-boris commented:
That is weird. It can also be a bug in the commons module because you have an incorrect state.
Likely it would happen when you trying to run next test after failed one. And the failed test didn't properly clean up suite state.

just-boris commented:
Well, the reason actually in the `allure-js-commons`: https://github.com/allure-framework/allure-js-commons/issues/17 I will fix it and you will not see that error.

But there can be bigger issue. I could only reproduce it with that code

```
beforeEach(() => {
   browser.url('/')
   // not return promise, so Mocha doesn't wait it.
})

it('test', () => {
   return browser.clik('.somewhere')
})
```

These commands are run in parallel because I intentionally forgot return in `beforeEach`. That causes race condition and it is probably bug. @gronke please check that your code don't have missed return statements.

gronke commented:
@just-boris great, thanks for the update. 

I filtered the log for the newly introduced warning message and deeply checked for missing promises in suites where it could occur, but could not find any mistakes like that. My feeling says it's somehow related to concurrency of tests. It'll take a while to run with different configurations and investigate. Will let you know asap.

From time to time steps with empty status or stop attribute show up in the results, which causes Allure to exit with an error when generating the results.
```
<step start="1466200678256" stop="" status="">
```
I would expect all steps to have a known status and start/stop time.

What's your plan for the `is not whithin any running test` warning? The order in which the webdriverio BaseReporter fires events does not match with the order required by Allure. Both, the branch this one was forked from and the refactored version I still did not finish yet rely on buffering the stack of suites and tests to deal with after hooks. The only way to tell a suite was closed was by seeing the next one start.

just-boris commented:
Actually now I am not going to no anything with this warning. I need just gather more information about it.
As I understand you are getting a lot of warnings like that because you have `before()` black outside of any `describe()`. Maybe that means that I need to bring back your workaround with extra top-level suite., but also I am looking for another, maybe more common solution.

christian-bromann commented:
@just-boris why is this PR pointing to webdriverio:master? We can't have the tests there. They should live here.

christian-bromann commented:
oh nevermind it's the owner of the repository not the repository name

christian-bromann commented:
@just-boris @gronke what is the state of this effort? I am really looking forward to use this. Any chance I can help?

just-boris commented:
Just fixed [this comment](https://github.com/webdriverio/wdio-allure-reporter/pull/17#issuecomment-226137700)

Don't know what to do about harmony flags [in that comment](https://github.com/webdriverio/wdio-allure-reporter/pull/17#issuecomment-226117496). But I hope we can solve it later since nothing has changed here.

Also, I introduced new warning message: https://github.com/just-boris/wdio-allure-reporter/blob/all-custom-reporter-events/lib/reporter.js#L75 It is very useful warning that means you got something wrong in your test setup, but it also logs `POST /wd/hub/session` that users can't resolve. I didn't find a way filter this event out from `runner:command` listener. So now it is going to bother users at start for each test session.

christian-bromann commented:
> if you okay with that caveats above, we can merge it and give a try for more users.

I am absolutely fine with that. Good job 👍

gronke commented:
Beautiful, I'm happy to see this merged!

just-boris commented:
@christian-bromann I don't have enough rights to perform the release. Could you make this? I think it should be released as `0.1.0` since did a lot of changes.

christian-bromann commented:
@just-boris yes absolutely. Thank you both for that effort!

christian-bromann commented:
@just-boris can run the test, see #18
